lib/smart_name.rb

```diff
@@ -44,20 +60,31 @@ def banned_re
   #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   #~~~~~~~~~~~~~~~~~~~~~~ INSTANCE ~~~~~~~~~~~~~~~~~~~~~~~~~
 
-  attr_reader :simple, :parts, :key, :s
+  attr_reader :namespaced, :simple, :rootspace, :parts, :key, :s
   alias to_s s
 
   def initialize str
     @s = str.to_s.strip
-    @s = @s.encode('UTF-8') if RUBYENCODING
+    @s = @s.encode('UTF-8')
     @key = if @s.index(self.class.joint)
         @parts = @s.split(/\s*#{JOINT_RE}\s*/)
         @parts << '' if @s[-1,1] == self.class.joint
         @simple = false
-        @parts.map { |p| p.to_name.key } * self.class.joint
+        names = @parts.each_with_object([]) do |part, names|
+            partial_key= names.flatten.map{ |p| p.to_name.key } * self.class.joint
+            if names.length == 0 || SmartName[partial_key]
```

Comment:
I guess this can be self.class[], but that only matters if [] might be overriden, and it probably should not be.